src/server/io.rs

```diff
@@ -200,3 +207,110 @@ impl Output for StdioOutput {
         RequestId::Num(self.next_id.fetch_add(1, Ordering::SeqCst) as u64)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn read_message_returns_message_from_valid_lsr_input() {
+        let mut input = io::Cursor::new("Content-Length: 7\r\n\r\nMessage");
+
+        let message =
+            read_message(&mut input).expect("Reading a message from valid input should succeed");
+
+        assert_eq!(message, "Message")
+    }
+
+    #[test]
+    fn read_message_fails_on_empty_input() {
+        let mut input = io::Cursor::new("");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Empty input should cause failure"
+        );
+    }
+
+    #[test]
+    fn read_message_returns_message_from_input_with_multiple_headers() {
+        let mut input =
+            io::Cursor::new("Content-Encoding: utf8\r\nContent-Length: 12\r\n\r\nSome Message");
+
+        let message =
+            read_message(&mut input).expect("Reading a message from valid input should succeed");
+
+        assert_eq!(message, "Some Message")
+    }
+
+    #[test]
+    fn read_message_returns_message_from_input_with_unknown_headers() {
+        let mut input =
+            io::Cursor::new("Unknown-Header: value\r\nContent-Length: 12\r\n\r\nSome Message");
+
+        let message =
+            read_message(&mut input).expect("Reading a message from valid input should succeed");
+
+        assert_eq!(message, "Some Message")
+    }
+
+    #[test]
+    fn read_message_fails_when_length_header_is_missing() {
+        let mut input = io::Cursor::new("Content-Encoding: utf8\r\n\r\nSome Message");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Reading a message with no length header should fail."
+        );
+    }
+
+    #[test]
+    fn read_message_fails_when_header_line_is_invalid() {
+        let mut input = io::Cursor::new("Invalid-Header\r\nContent-Length: 12\r\n\r\nSome Message");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Reading a message with invalid header should fail."
+        );
+    }
+
+    #[test]
+    fn read_message_fails_when_length_is_not_numeric() {
+        let mut input = io::Cursor::new("Content-Length: abcd\r\n\r\nMessage");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Reading a message with no length header should fail."
+        );
+    }
+
+    #[test]
+    fn read_message_fails_when_length_is_too_large_integer() {
+        let mut input = io::Cursor::new("Content-Length: 1000000000000000000000\r\n\r\nMessage");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Reading a message with no length header should fail."
```

Comment:
nit: this should mention length being to large, correct?

Reply:
Yes, sorry about it.